github_workflow.md

```diff
@@ -0,0 +1,153 @@
+# GitHub Milestone and Issue Workflow
+
+This purpose of this document is to describe our development workflow on
+GitHub; in particular, how we use milestones and issues. We recognize that our
+many subprojects may use variations of these ideas. This particular workflow is
+based on the practices from the JupyterLab repo
+(https://github.com/jupyterlab/jupyterlab) and are optional for Jupyter
+subprojects to follow.
+
+## Milestones
+
+We use Github milestones to track issues and pull requests related to particular
+releases of a subproject (https://github.com/jupyterlab/jupyterlab/milestones).
+
+* These milestones should be labeled with the version numbers of the releases
+  (`4.3`, `5.0`, etc.).
+* Subprojects should only have a few numbered milestones. Beyond one or two
+  future releases, we have found that detailed milestones become meaningless.
+* Subprojects should also have a `Backlog` milestone for issues and pull
+  requests that have not been assigned to a particular release milestone. This
+  also includes new feature requests ("wishlist") issues that we haven't
+  committed to implementing. Essentially, `Backlog` means "at some point in the
+  future, maybe."
+* Subprojects should have a `No Action` milestone to indicate issues or pull
+  requests that were closed with no action or work being performed. In general
+  we try to close issues/PRs with this milestone.
+* The only time an issue should have no milestone is early on in the triage
+  process when it still has the `status:Needs Info` label.
+
+## Labels
+
+We use GitHub labels to organize issues and pull requests
+(https://github.com/jupyterlab/jupyterlab/labels). The Github issues namespace
+is completely unstructured; each label is simply a name with an associated
+color. This unstructured nature, makes it difficult to organize work
+effectively. To address, this challenge, we are started to use a more formal
+label schema/taxonomy.
+
+First, every label should have one of four prefixes:
+
+1. `type:`
+2. `component:`
+3. `cat:`
+4. `status:`
+
+All issues of a given prefix should have the same color to make visual
+identification easy.
+
+### `type:`
+
+The `type:` prefix indicates the type of issue or pull request. There are four
+types:
+
+* `type:bug`
+* `type:enhancement`
+* `type:maintenance`
```

Comment:
What is the practical effect of distinguishing between bug/enhancement/maintenance work to do, though?  i.e. is anything different in terms of directing attention based on these classifications? If I see 5 bugs and 10 enhancements open for a milestone, is anything different about what I should do today? What about an aspiring volunteer? When we had bug/enhancement labels the first time, the answer (for me, at least) was definitely "No." I *did*, however, have to spend a lot of time going through issues, trying to decide if bug or enhancement was the right name to use, which ultimately had no apparent effect on anybody else's attention or work. It seems to still be tasks of the same kind (writing code / docs) that need to be accomplished in the same time frame. The useful piece seems to be 'action to take' (bug, enhancement, maintenance) vs `type:question`, which is a discussion to resolve, rather than a PR to make (also corresponds to `milestone: no action`).

We still have the 'bug' label on most of our repos, and I find that it often gets applied to issues people feel are 'important', even if they are changes in behavior that I would consider an 'enhancement', albeit an important and/or small one. There is no disagreement about whether, when, or how the issue should be addressed, but the addition of the bug/enhancement label introduces a source of disagreement whose resolution doesn't seem to affect anything. I'd prefer to avoid having discussions about "I disagree that this should be called a bug, but I agree that it should still change", if we can.